File: baekjoon/basic3/16234/16234.cpp
// 백준 16234 : 인구 이동 (https://www.acmicpc.net/problem/16234)
#include <iostream>
#include <vector>
#include <algorithm>
#include <cstdlib>
using namespace std;
int n, l, r, ret, ny, nx, sum;
bool flag;
vector<pair<int, int>> v;
int a[54][54], visited[54][54];
int dy[] = {0,0,1,-1};
int dx[] = {1,-1,0,0};

void dfs(int y, int x){
    sum += a[y][x];

    for(int i=0; i<4; i++){
        ny = y + dy[i];
        nx = x + dx[i];
        if(ny >= n || ny < 0 || nx >= n || nx < 0 || visited[ny][nx]) continue;
        if(abs(a[ny][nx] - a[y][x]) >= l && abs(a[ny][nx] - a[y][x]) <= r){
            visited[ny][nx] = 1;
            v.push_back({ny, nx});
            dfs(ny, nx);
        }
    }
    return;
}

int main(){
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);

    cin >> n >> l >> r;
    for(int i=0; i<n; i++){
        for(int j=0; j<n; j++){
            cin >> a[i][j];
        }
    }

    int day = 0;
    while (true){
        flag = 0;
        fill(&visited[0][0], &visited[0][0] + 54*54, 0);
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                v.clear();

                sum = 0;
                if(visited[i][j]) continue;
                v.push_back({i, j});
                visited[i][j] = 1;
                dfs(i, j);
                if (v.size() == 1) continue;

                for(pair<int, int> p : v){
                    a[p.first][p.second] = sum / v.size();
                    flag = 1;
                }
            }
        }
        if(!flag) break;    
        day++;
    }
    
    cout << day << "\n";
    
    return 0;
}

File: baekjoon/basic3/16234/16234.py
# 백준 16234 : 인구 이동 (https://www.acmicpc.net/problem/16234)
import sys
sys.setrecursionlimit(10**6)

dy = [0,0,1,-1]
dx = [1,-1,0,0]

def go(y, x):
    global is_move
    
    for d in range(4):
        ny = y + dy[d]
        nx = x + dx[d]
        if ny >= n or ny < 0 or nx >= n or nx < 0:
            continue
        if abs(a[y][x]-a[ny][nx]) >= l and abs(a[y][x]-a[ny][nx]) <= r:
            line[y][x].append((ny, nx))
            is_move = True
    return

def dfs(y, x, group):
    visited[y][x] = group
    person_cnt = a[y][x]
    kan_cnt = 1

    for v in line[y][x]:
        ny = v[0]
        nx = v[1]
        if visited[ny][nx] == group:
            continue

        ret1, ret2 = dfs(ny, nx, group)
        person_cnt += ret1
        kan_cnt += ret2
    return person_cnt, kan_cnt
        
n, l, r = map(int, sys.stdin.readline().split())

a = []
for i in range(n):
    a.append(list(map(int, sys.stdin.readline().split())))

day = 0
while True:

    is_move = False

    # 국경선 초기화, 연합 초기화
    line = [[[] for _ in range(n)] for _ in range(n)]
    visited = [[0] * n for _ in range(n)]

    # 국경선 open
    for i in range(n):
        for j in range(n):
            go(i, j)
    
    # 연합
    group = 1
    group_person = [-1]

    for i in range(n):
        for j in range(n):
            if visited[i][j] != 0:
                continue
            person_cnt, kan_cnt = dfs(i, j, group)
            person = person_cnt // kan_cnt
            group_person.append(person)
            group += 1
    
    for i in range(n):
        for j in range(n):
            if visited[i][j] == 0:
                continue
            a[i][j] = group_person[visited[i][j]]
    
    if is_move == False:
        break

    day += 1

sys.stdout.write(str(day)+"\n")